sTGC offline reconstruction chain / makers

jdbrice opened the pull request:
The majority of the sTGC offline reconstruction chain:
- Updates to StEvent/StFtt*
- Updates to StFttRawHitMaker to accommodate new data members
- StFttHitCalibMaker - calibrated the bunch crossing id of hits
- StFttClusterMaker : Finds clusters of hits via TPC-like algorithm
- StFttDbMaker and StFttDb : central store of parameters and conversion functions. Currently reads from TXT or data members, NO db access
- StFttPointMaker : Makes local and global spacepoints from clusters

Needed for fast offline QA package which will come in the next PR

Test chain on recent DAQ file:
root4star -b -q -l 'bfc.C( 10000, "in, db, StEvent, MuDST, ftt, FttQA", "input.daq", "TEST.root" )'

eg. 
input.daq ->
/star/data03/pwg/jdb/FWD/daq/st_physics_23019034_raw_0000001.daq

jdbrice committed:
Merge pull request #1 from star-bnl/main

update from upstream on Friday, June 11, 2021

jdbrice committed:
Initial readme on how to sparse checkout StarVMC

Commit:
add .gitignore for working directories StarGeometry and xgeometry

Commit:
edit description comment

Commit:
Rotated middle disk by 30 deg to mactch actual installation.

fvidebaek committed:
simplified base file for fst geometry
Updated StarGeo.xml to be able to acces this in dev2021

gavin-wilks committed:
Simplified FST geometry v1.1 ( cooling tubes, ribs, cooling tube holder, heatsinks (draft) )

gavin-wilks committed:
Adjustments to commented out simplified heatsinks and included 30 deg rotation of middle disk

gavin-wilks committed:
Fixed syntax issue with middle disk rotation

fvidebaek committed:
Updated material definition for combined stainless and novec

fvidebaek committed:
Merge branch 'main' of https://github.com/jdbrice/star-sw-1 into main
This includes material changes

gavin-wilks committed:
Added 3 more cooling tubes, compressed heatsinks into single volumes in PEEK bases, removed unnecessary code

gavin-wilks committed:
Compressed heatsinks and added 3 additional cooling tube volumes (simplified cooling tube connectors)

fvidebaek committed:
accept gavins changes. removed some dummy not used lines

fvidebaek committed:
Ring removed.

jdbrice committed:
remove README for pull request

plexoos committed:
Fixes for PR

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

Commit:
Write StMuFttXXX from StEvent equivalents

danbburg committed:
fix copy typo, FcsAll->FttAll

danbburg committed:
fix ROOT6 build failure, include iostream header in StMuFttRawHit.h

danbburg committed:
Initial stgc-offline chain with working cluster finder and chain options

danbburg committed:
Update correct mapping for diagonal, QA for clusters, fix clustre maker

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

jdbrice committed:
Merge branch 'star-bnl:main' into main

Commit:
updates for StFtt* for cluster maker and offline QA

Commit:
updates for StFtt* for cluster maker and offline QA

Commit:
resolve merge from main in StMuDst/COMMON/StMuFtt*

jdbrice committed:
Merge branch 'star-bnl:main' into main

Commit:
merge in main and resolve conflicts with StBFChain

Commit:
cleanup, comments, prints, add fttChain

Commit:
fix typo and build BFChain docs

Commit:
fix typo in fttQA chain option

klendathu2k commented on StRoot/StBFChain/doc/StBFChain.cxx_doc:
```diff
@@ -1271,6 +1271,7 @@ Bfc_st BFC[] = {
   {"eemcDb" ,"eeDb" ,"","db,EEmcUtil", "StEEmcDbMaker","StEEmcDbMaker","Load EEmcDbMaker",kFALSE},
```
Aren't the files in the "doc" directly are produced as a build artifact?

genevb replied:
Right, these files should be excluded from the PR.

klendathu2k replied:
I see you approved the full PR, including these changes.... so I'll withhold my approval until this is resolved.

genevb replied:
Yeah, I probably should have just submitted mine as a comment instead of approval. One of many complications in life with git. Sorry.

klendathu2k commented on StRoot/StEvent/StFttPoint.h:
```diff
@@ -57,14 +57,14 @@ inline UChar_t StFttPoint::plane() const { return mPlane; }
 inline float StFttPoint::y() const { return mY; } // y position (cm) in local coords.
-inline int StFttPoint::nParentClusters() const { return mClusters.size(); } // Number of points in parent cluster
-inline StFttCluster* StFttPoint::cluster( size_t i ) { return mClusters[i]; } //  Parent cluster of the photon.
+// inline int StFttPoint::nParentClusters() const { return mClusters.size(); } // Number of points in parent cluster
```
Remove commented-out-code.

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StEvent/StFttPoint.h:
```diff
@@ -57,14 +57,14 @@ inline UChar_t StFttPoint::plane() const { return mPlane; }
-inline int StFttPoint::nParentClusters() const { return mClusters.size(); } // Number of points in parent cluster
-inline StFttCluster* StFttPoint::cluster( size_t i ) { return mClusters[i]; } //  Parent cluster of the photon.
+// inline int StFttPoint::nParentClusters() const { return mClusters.size(); } // Number of points in parent cluster
+inline StFttCluster* StFttPoint::cluster( size_t i ) { return mClusters[i % 4]; } //  Parent cluster of the photon.
```
Taking i%4 is dangerous.  If a user makes an off-by-one error in the code which calls it (i.e. feeds in a range of values [1,4] instead of the expected [0,3]), the code will silently work, and return unexpected results.  This type of bug will be difficult to track down.

This should return a nullptr when a value outside of the legal range is given.  Similarly, addCluster should perform no action if an illegal index is passed in.

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.h:
```diff
@@ -0,0 +1,65 @@
+/***************************************************************************
+ *
+ * $Id: StFttClusterMaker.h,v 1.3 2019/02/19 20:32:05 fseck Exp $
```
$Id$ is leftover from CVS days.   Can be removed.

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.h:
```diff
@@ -0,0 +1,65 @@
+    ~StFttClusterMaker();
+
+
+    Int_t  Init();
```
Coding standards prefer basic C++ types (int, double, char, etc...) over ROOT types (Int_t, Double_t, Char_t, etc...) for non-persistent classes.

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttPointMaker/StFttPointMaker.cxx:
```diff
@@ -0,0 +1,214 @@
+Int_t
+StFttPointMaker::Make()
+{ 
+    LOG_INFO << "StFttPointMaker::Make()" << endm;
```
No need to increase the size of the log file here...

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttPointMaker/StFttPointMaker.cxx:
```diff
@@ -0,0 +1,214 @@
+void StFttPointMaker::MakeGlobalPoints() {
+    for ( StFttPoint * p : mFttCollection->points() ){
+
+        double z = p->plane() * 60;
```
Global transformation is hardcoded... what is the plan to apply alignment measurements?

jdbrice replied:
sTGC is a bit behind on the DB side, so we have no actual databases setup yet. But your point is taken, we can at least setup the infrastructure to query the StFttDbMaker for these values instead of hard coding.

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

jdbrice commented:
Note: Based on the logs, the builds are failing because RTS is out of date. Even though the relevant changes were made a while ago, I thought it was auto-updated CVS->Git ever so often?

genevb reviewed (approved):
The BFC modifications look reasonable to me.

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -0,0 +1,429 @@
+/***************************************************************************
```
This is a new maker.  Review for this will be a little more detailed.

jdbrice replied:
Right, two reviewers have been identified (one agreed, one waiting). But your comments are appreciated nevertheless.

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -0,0 +1,429 @@
+    if(mEvent) {
+        LOG_DEBUG<<"Found StEvent"<<endm;
+    } else {
+        return kStOk;
```
"return kStWarn" would be a bit more appropriate.  Also, it makes a bit more sense to issue a DEBUG message on the exceptional case (StEvent missing) rather than the case where everything is working as expected.

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -0,0 +1,429 @@
+
+    mFttDb = static_cast<StFttDb*>(GetDataSet("fttDb"));
+
+    LOG_INFO << "Found " << mFttCollection->rawHits().size() << " Ftt Hits" << endm;
```
Probably do not need to output this on every event.  Suggest LOG_DEBUG instead.  (And perhaps only if no hits found).

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -0,0 +1,429 @@
+
+
+bool StFttClusterMaker::PassTimeCut( StFttRawHit * hit ){
+    return (abs( hit->time() ) <=3); 
```
Useful to make this cut value a DB parameter as well?

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -0,0 +1,429 @@
+                             [](const StFttRawHit* a,const StFttRawHit* b) { return a->adc() < b->adc(); });
+
+    pos = (itMax - hits.begin());
+    if ( pos >= hits.size() )
```
Why not just use ...

if ( itMax == hits.end() ) return nullptr;

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -0,0 +1,429 @@
+    StFttRawHit *hitLeft = nullptr, *hitRight = nullptr;
+
+    while ( searchRight || searchLeft ){
+        if ( mDebug ){
```
--> LOG_DEBUG

klendathu2k replied:
All classes which inherit from StMaker have a SetDebug( int debuglvl ) method.  You can retrieve the debug level using the Debug() method.  I believe that the logger refers to the debug level in order to decide whether LOG_DEBUG lines should be written or not.

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k commented on StRoot/StFttClusterMaker/StFttClusterMaker.cxx:
```diff
@@ -0,0 +1,429 @@
+        // Now find the cluster edges
+        size_t left = anchor, right = anchor;
+        SearchClusterEdges( hits, anchor, left, right);
+        if ( mDebug ){
```
Again --> LOG_DEBUG

jdbrice replied:
Addressed: 3686a6eea3b300c95180bdcee0ad698b5763a88c

klendathu2k reviewed (dismissed):
I don't see any show stoppers with the code.  A few changes are requested and/or suggestions noted.

akioogawa commented on StRoot/StFttDbMaker/StFttDb.cxx:
```diff
@@ -0,0 +1,265 @@
+ * \author: jdb
+ ***************************************************************************
+ *
+ * Description: This interface between FCS and the STAR database
```
FCS?

jdbrice replied:
Addressed in 3686a6eea3b300c95180bdcee0ad698b5763a88c

akioogawa commented on StRoot/StFttDbMaker/StFttDb.h:
```diff
@@ -0,0 +1,133 @@
+ * \author: akio ogawa
+ ***************************************************************************
+ *
+ * Description: FCS DB Utility
```
FCS?

jdbrice replied:
Addressed in 3686a6eea3b300c95180bdcee0ad698b5763a88c

akioogawa commented on StRoot/StFttDbMaker/StFttDb.h:
```diff
@@ -0,0 +1,133 @@
+/***************************************************************************
+ * $Id: StFttDb.h,v 1.3 2021/05/27 14:02:23 akio Exp $
+ * \author: akio ogawa
```
Not me!

jdbrice replied:
Addressed in 3686a6eea3b300c95180bdcee0ad698b5763a88c

akioogawa commented on StRoot/StFttDbMaker/StFttDbMaker.cxx:
```diff
@@ -0,0 +1,46 @@
+ * \author: jdb
+ ***************************************************************************
+ *
+ * Description: This maker is the interface between FCS and the STAR database
```
FCS?

jdbrice replied:
Addressed in 3686a6eea3b300c95180bdcee0ad698b5763a88c

Commit:
address jason's comments

starsdong commented:
Daniel, thank you very much for the update and addressing these comments? There is one outstanding issue related to the StFttDbMaker. Currently it still uses local file. Is the intention still to request DB idl tables and go through the standard route?

jdbrice commented:
> Daniel, thank you very much for the update and addressing these comments? There is one outstanding issue related to the StFttDbMaker. Currently it still uses local file. Is the intention still to request DB idl tables and go through the standard route?

Xin, i have prepared the DB items and will send to Dmitri today asap. After that is ready I'll update the DbMaker to use the database. 
We will need another re-review wen all of the other databases are integrated in the future

starsdong commented:
Daniel, we have the new DB idl files created. Do I understand correctly that you will be updating the FttDbMaker accordingly and update this PR? Thanks

Commit:
integrate Geometry/ftt/fttHardwareMap and Calibrations/ftt/fttDataWindows

jdbrice commented:
@starsdong and @dmarkh I've updated the pull request to address the databases that are currently in use. Thank you Dmitry for taking care of those. If the DB updates checkout then we should be ready to merge this PR

starsdong commented:
Dmitry A. has reviewed the new commit and OKed with the code changes. One comment from Dmitry:
""fttHardWarMap" table is created as a non-indexed table. It won't be possible to convert "non-indexed" table to "indexed" - it won't be backward-compatible."

With this, I think we can conclude the code review and merge this PR. Thanks to everyone who helped on this.

starsdong commented:
Dmitri, one technical question before merging. Some directories (StBFChain) in the main branch have received new commits since last time Daniel merged the main to his branch. Is it still OK that we merge the current PR into main directly now?

plexoos commented:
It should be OK to merge. Otherwise, I would expect a warning about conflicting changes displayed on the PR page.

starsdong commented:
Thanks, Dmitri. I will click the merge then.
